<a href="https://colab.research.google.com/github/dwaidatta/ds-ai-nfsu/blob/main/ColabNotebooks/Main_Course_Resource_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 - Introduction to Python

This notebook covers the coding aspects of python programming language.

## A new day in the programming world!

We are in the world of programming. This is where we give instructions to a computer and tell it what to do. How do we give an instruction to a computer? Through a program! We write programs which the computer executes to perform the task we want.

The first thing that we do is say hello to our new world!

In [ ]:
print("Hello World!")

Hello World!


Well, you just said hello!

Here `print()` is a function and inside it is a **string**. The string is enclosed in double quotes `" "`. This is the syntax to say/show something. Print function helps us to output things on the screen.

Let's move on.

Here we will see a comment. A comment helps us to document our code or add comments. A comment does not affect the execution or logic of the code. A comment starts with the `#` symbol.

In [ ]:
# This is a comment

Before we move on, there are some important things that we should discuss. Python programs/scripts have the `.py` file extension. The one you are seeing here is a python jupyter notebook. Well no need to worry much. It runs python scripts/programs too. You just have more control running code snippts back and forth and making changes to the code if required without starting to run all the lines of code from scratch.

This notebook has a `.ipynb` file extension.

*Another thing which you may come across various cells below, is the use of `%%python`, `%%python3`, etc. **Do NOT** be worried about them for your course or consider them while writing your code. These are special magic tricks that help each cell to behave as independent program and the variables/functions are not carried over to the next cell. This helps me demostrate each code snippet individually better.*

Let's move on.

Suppose we want the user to give his name to us, and we want to greet him. Lets try to code that.

In [ ]:
name = input("Enter your name: ")
print("Hello", name)

Enter your name: Alice
Hello Alice


Woah! That's a lot of code all together. Let us understand step by step.

First, you should understand that each line by itself is a individual statement. This statement executes by itself and has its own logic. We use two lines here:

- The first line uses the `input()` function to ask for an input value from the user. This is called **user input**. The user input is then stored into the `name` variable.
- The second line uses the `print()` function to say Hello with the name. How did the spaces come, whats the formatting logic? We will see them soon.

That's a good start. Let's move onto the next chapters to learn more.

## The Python Propaganda!

Python has become the go-to language for Data Science and AI, not by accident, but because of specific design choices and ecosystem advantages. This chapter explores the "why" behind Python's dominance.

### Why Python for Data Science?

Python is preferred in Data Science and AI for several structural reasons:

- Simple, readable syntax that resembles plain English, reducing the learning curve for non-programmers like statisticians and domain experts
- A vast ecosystem of specialized libraries (NumPy, Pandas, Matplotlib, Scikit-learn, TensorFlow, PyTorch) that handle everything from data manipulation to deep learning
- Strong community support, meaning most problems already have documented solutions or forum discussions
- Interpreted nature, allowing quick testing and iteration without a compile step
- Cross-platform compatibility, so the same code runs on Windows, macOS, and Linux
- Seamless integration with other languages (C, C++, Java) for performance-critical tasks

### Python vs Traditional Programming Languages

| Aspect | Python | Traditional Languages (C/Java) |
|---|---|---|
| Syntax complexity | Minimal, close to pseudocode | Verbose, strict typing rules |
| Development speed | Fast prototyping | Slower due to boilerplate code |
| Library support for ML/AI | Extensive and mature | Limited or requires more setup |
| Community and resources | Massive, active | Smaller in the AI/ML domain |
| Execution speed | Slower (interpreted) | Faster (compiled) |

### Role of Python in the AI/ML/Data Science Pipeline

- Data collection and web scraping (using libraries built for handling APIs and files)
- Data cleaning and preprocessing (removing noise, handling missing values)
- Exploratory data analysis (visual and statistical understanding of data)
- Model building and training (implementing ML/DL algorithms)
- Model evaluation and deployment (testing accuracy and pushing models to production)

### The Bigger Picture

Python's simplicity does not mean it sacrifices power; its real strength lies in acting as a bridge between mathematical theory and practical implementation. This balance of accessibility and capability is what has made it the unofficial language of modern AI and Data Science.

## Understaning Python Basics

Let us come to the core python code and its syntax. Every programming language, just like english language, has its grammar or a ruleset to follow while we write it.

Think of it as a person coming upto you and saying: "School late car was because my I was reach for".

Did you understand what he meant? No, right? There is a specific way he should arrage the words, subject and predicate and use proper tenses so that you are able to understand what he is saying.

In the same manner, programming languages have their own rules to help the computer understand what the program should do. Deviating from that ruleset/syntax breaks the code and the program runs into an error.

*That previous text was actually: I was late for school because my car was late.*

### Input Output

The `print()` function is used for output and the output can be formatted as per needs.

The `input()` function takes input from the user.

In [ ]:
print("Hello")
print("100")
print(100)

Hello
100
100


Here on the above cell the print function prints several values. Let's understand them.

- The first print statement prints the `Hello` string.
- The second print statement prints the `100` string.
- The third print statement prints the `integer 100`.

But we need not distinguish the type of the output. This is merely to show that print can accept more than just strings.

*String is a sequence of characters, mainly alphanumeric and other symbols.*

Let's see another example.

In [ ]:
print("Hello", "Alice")

Hello Alice


Here the `print` function takes two values inside itself separated by a `comma`. The comma acts as a separator and gives a `space` **by default** between the values given.

In [ ]:
print()

An empty print statement just prints a blank line.

Now, let's see inputs.

In [ ]:
input("Enter name: ")

Enter name: Alice


'Alice'

When this statement is executed, the user is prompted with `Enter name: `. The user then inputs the values which is shown.

*Note the quotes around the word. This means that by default input function returns the user input in form of a string.*

*Also note that the `'Alice'` is actually not an output. It is a feature of the python notebook to show the input taken. Follow the below cells to understand how it behaves in a normal python script.*

The input function can be used without any prompt inside it too. For example:

In [ ]:
input()

'Alice'

If you get confused how these both are different because you cannot see the prompt being shown, I will put in two examples from `.py` script.

So

```python
input("Enter name: ")
```

executes as

```text
Enter name: Alice
```

Whereas,

```python
input()
```

executes as

```text
Alice
```

We gave input of the string value `Alice` in both cases.

### Variables

Now, that you know how to take input from the user and show things on the screen, you will want to go a step further.

Where does the values go which we take input from the user?

Let us say there is an app where you register using your name and age. The name and age needs to be stored for later use too. There comes the concept of variables.

Variables are named memory locations to hold different kind of values.

A named memory location means that there is some space on the computer memory hardware and we are using a name to use that space to store our specified value.

Let's see some examples.


In [ ]:
var1 = 10

Here, we store the value `10` inside of `var1`. That's how we say it.

So now anywhere if we use `var1`, we should know that it holds a value of `10` until changed or deleted.

Let's try to print `var1`.

In [ ]:
print(var1)

10


10 is printed on the output. Here the print function takes the `var1` named variable inside it (as a parameter).

In [ ]:
var1 = 10

So, you may already understand the syntax. The right-side value is stored into the left-side variable. The `=` sign is an **assignment operater**. This means the `=` sign assigns the value `10` to the variable `var1`.

### Datatypes - Part 1

Before going any further, we will discuss few important things. One very important thing is datatype.

Every value that we use in python or any other programming language has a datatype.

Datatype means the type of data - whether it is an number, a string, or a boolean value etc. Based on the type of data, we have different valid operations or functions that we can perform on that data.

Let's try to understand that you have 10 apples. Here the number 10 is an integer. Now, your friend has 20 apples. Here 20 is an integer too. Certainly you are able to add 10 and 20, and get 30 apples. You were able to add those numbers.

Now, if your name is Alice and your friend's name is Kara, are you able to sum those two names? No!

Another example is that, a question is asked to you and your friend that whether you have completed the assignments or not. You say `Yes` and your friend says `No`. What can be the valid operation on such values of yes and no?

Let us see the basic ones for now. Try to follow the names as given here. There are named as so in python and we have to follow that.

For numeric datatypes, we generally have:

- Integer: Denoted as `int`. Integer values. Example: 10, -20, 30, 0 etc

- Float: Denoted as `float`. Numeric values with the dot! Example: 10.1, -23.0, 3.14, 2.22 etc

- Complex Numbers also exist but very little application in this course.

For alphanumerics and other symbols we have:

- Character: One single symbol by itself. Example: A, B, c, d, 1, #, @ etc. (All of these needs to be enclosed in quotes.)

- String: Denoted by `str`. Sequence of character(s). Example: ABcd, A#2124, Alice, 132324 etc. (All of these needs to be enclosed in quotes.)



### Connecting data, datatypes and variables

We were using various data till now. We took input and used print. We stored values in variables too. It is important to understand that every value has its datatype and therefore its valid operations too.

Let's understand with examples now.

In [ ]:
var1 = 10

The 10 here is of the type integer and valid operations are as per its datatype (we will discuss that later).

In [ ]:
var1 = 10.0

The 10 here is written as 10.0 which means we are using the float datatype and its a numeric 10.0

In [ ]:
var1 = "10"
var2 = "10.0"

Here both the variables have been assigned two values which are strings since they are enclosed in quotes. We will not worry about any numeric similarity here. For programming, the strings `"10"` and `"10.0"` are just separate values.

The important fact is that in Python, a programmer (you) do not need to **explicitly** mention the datatype of a variable which needs to be done in other programming languages.

Infact, in Python, we have **dynamic assignment**, which means that we assign an integer to a variable at some point, and later assign some value of different datatype without any issue. Let's see an example.

In [ ]:
var1 = 100
print("Value is: ", var1)
print("Type is: ", type(var1))

# reassign different datatype value now

var1 = "Alice"
print("Value is: ", var1)
print("Type is: ", type(var1))

Value is:  100
Type is:  <class 'int'>
Value is:  Alice
Type is:  <class 'str'>


From the output it should be clear that `var1` had stored an interger 100 before. The `type()` function confirms the datatype too.

For the 2nd time, we have assigned a string `"Alice"` to `var1`. The datatype the variable is automatically adjusted by python internally. Also we can confirm the string datatype using the `type()` function.

Also at this point it should be clear, that changing the value of a variable is definitely possible since the name itself is variable. So the below example should be naive to include here:

In [ ]:
var1 = 10
print(var1)

# change value

var1 = 20
print(var1)

10
20


Now, let's understand that user input can be stored into variables. Let's try that with some examples and explanations.

In [ ]:
name = input("Enter your name: ")
age = input("Enter your age: ")

# let us print the name and age to see if we have the values saved

print("Name:", name)
print("Age:", age)

Name: Alice
Age: 20


Well that works! But let's run the program again and see the datatypes of them once. As said earlier, input function takes inputs in string format by default. So both of them should be strings.

In [ ]:
name = input("Enter your name: ")
age = input("Enter your age: ")

# let us print the name and age to see if we have the values saved

print("Name:", name)
print("Type of name:", type(name))
print("Age:", age)
print("Type of age:", type(age))

Name: Alice
Type of name: <class 'str'>
Age: 20
Type of age: <class 'str'>


Well, both are in string. But certainly you will want the name to be a string, and the age to be of an integer type (don't bother about negative or zero age). Let's try to do that and understand via code.

In [ ]:
name = input("Enter your name: ")
age = int(input("Enter your age: "))

# let us print the name and age to see if we have the values saved

print("Name:", name)
print("Type of name:", type(name))
print("Age:", age)
print("Type of age:", type(age))

Name: Alice
Type of name: <class 'str'>
Age: 20
Type of age: <class 'int'>


We used something called **explicit type casting** here. It means changing the datatype of something into a datatype that we want. Since a string `20` has a valid integer representation, we can convert it without issues. But remember the same cannot be done with some value like `Alice`.

The `int(...)` typecasted the string 20 into integer value 20 which is confirmed from the type output.

We also need to keep in mind, few rules while naming variables. They are:

- Can contain letters, digits and underscores (_)
- First character cannot be a digit
- Keywords cannot be used as variable names

**Keywords** are special words that are reserved by the python programming language. Every programming language has its own reserved words. In python few keywords are:
`if else True False class int float None in not` and many more.

Also the naming is **case-sensitive**, meaning that `Name` and `name` will be treated as two different variables.

Let's see a few examples to understand these better.

In [ ]:
var_1 = 10

The above naming is valid.

In [ ]:
1_var = 10

SyntaxError: invalid decimal literal (3941448491.py, line 1)

This naming is invalid and we get an error since the variable name starts with a number.

In [ ]:
_var1 = 10

This naming is valid too.

In [ ]:
class = 10

SyntaxError: invalid syntax (3305041997.py, line 1)

This naming is invalid since it is a keyword.

In [ ]:
class123 = 10

This name is valid since `class123` is not a keyword.

Now let's see the **case-sensitive** naming example.

In [ ]:
name = "Kara"
Name = "Alice"
nAme = "Luther"
NAME = "Jerry"

print(name)
print(Name)
print(nAme)
print(NAME)

Kara
Alice
Luther
Jerry


From the above output, it should be clear that the variables clearly store different values. If they were **case-INsensitive**, then they would have printed the last value since that would be considered a case of variable value update.

### Operators

We do operations in day to day life, life adding and subtracting. That is called an operation. Now say in `20 + 10`, the `+` sign or symbol is the operator and the values `20` and `10` are called values.

In python too, we have operators and operands. Operators are special symbols that operate on operands. Operands are the values on which an operator acts. Operations have a specific outcome.

Let's understand with examples.

In [ ]:
total = 10 + 20
print(total)

30


Here, we have used the addition operator `+`. It operates on two operands(values) `10` and `20`. The result 30 is stored in the total variable.

#### Arithmetic Operators
Used to perform basic mathematical calculations.

| Operator | Name | Description | Example | Result |
|---|---|---|---|---|
| `+` | Addition | Adds two values | `5 + 2` | `7` |
| `-` | Subtraction | Subtracts right value from left | `5 - 2` | `3` |
| `*` | Multiplication | Multiplies two values | `5 * 2` | `10` |
| `/` | Division | Divides and always returns a decimal (float) | `5 / 2` | `2.5` |
| `//` | Floor Division | Divides and rounds down to nearest whole number | `5 // 2` | `2` |
| `%` | Modulus | Gives the remainder after division | `5 % 2` | `1` |
| `**` | Exponent | Raises left value to the power of right | `5 ** 2` | `25` |

---

#### Assignment Operators
Used to assign values to variables, often combined with arithmetic for shortcuts.

| Operator | Description | Example | Same As |
|---|---|---|---|
| `=` | Assigns value to a variable | `x = 5` | — |
| `+=` | Adds and reassigns | `x += 2` | `x = x + 2` |
| `-=` | Subtracts and reassigns | `x -= 2` | `x = x - 2` |
| `*=` | Multiplies and reassigns | `x *= 2` | `x = x * 2` |
| `/=` | Divides and reassigns | `x /= 2` | `x = x / 2` |
| `//=` | Floor divides and reassigns | `x //= 2` | `x = x // 2` |
| `%=` | Modulus and reassigns | `x %= 2` | `x = x % 2` |
| `**=` | Exponent and reassigns | `x **= 2` | `x = x ** 2` |

---

#### Comparison (Relational) Operators
Used to compare two values. Always returns a Boolean (`True` or `False`).

| Operator | Meaning | Example | Result |
|---|---|---|---|
| `==` | Equal to | `5 == 5` | `True` |
| `!=` | Not equal to | `5 != 3` | `True` |
| `>` | Greater than | `5 > 3` | `True` |
| `<` | Less than | `5 < 3` | `False` |
| `>=` | Greater than or equal to | `5 >= 5` | `True` |
| `<=` | Less than or equal to | `3 <= 5` | `True` |

---

#### Logical Operators
Used to combine multiple conditions. Common in `if` statements and loops.

| Operator | Meaning | Example | Result |
|---|---|---|---|
| `and` | True only if BOTH conditions are true | `5 > 2 and 3 > 1` | `True` |
| `or` | True if AT LEAST ONE condition is true | `5 > 10 or 3 > 1` | `True` |
| `not` | Reverses the result (True becomes False, and vice versa) | `not(5 > 2)` | `False` |

---

#### Membership Operators
Used to check whether a value exists inside a sequence like a list, string, or dictionary.

| Operator | Meaning | Example | Result |
|---|---|---|---|
| `in` | True if value is found in the sequence | `3 in [1, 2, 3]` | `True` |
| `not in` | True if value is NOT found in the sequence | `4 not in [1, 2, 3]` | `True` |

---

#### Identity Operators
Used to check if two variables point to the exact same object in memory (not just equal in value).

| Operator | Meaning | Example |
|---|---|---|
| `is` | True if both variables refer to the same object | `x is y` |
| `is not` | True if variables refer to different objects | `x is not y` |

---

### Operator Precedence and Associativity

**Precedence** decides which operator is evaluated first when multiple operators appear in one expression. **Associativity** decides the evaluation direction (left-to-right or right-to-left) when operators have the same precedence.

| Precedence (High → Low) | Operator(s) | Associativity | Notes |
|---|---|---|---|
| 1 | `()` | Left to right | Parentheses evaluated first, always |
| 2 | `**` | Right to left | Exponent is evaluated right-to-left |
| 3 | `+x`, `-x` (unary) | Right to left | Unary plus/minus, e.g. `-5` |
| 4 | `*`, `/`, `//`, `%` | Left to right | Multiplication group |
| 5 | `+`, `-` (binary) | Left to right | Addition/subtraction group |
| 6 | `==`, `!=`, `>`, `>=`, `<`, `<=` | Left to right | Comparison operators |
| 7 | `is`, `is not`, `in`, `not in` | Left to right | Identity and membership checks |
| 8 | `not` | Right to left | Logical NOT |
| 9 | `and` | Left to right | Logical AND |
| 10 | `or` | Left to right | Logical OR |

**Quick Example:** In the expression `2 + 3 * 2 ** 2`, Python evaluates `**` first (`2 ** 2 = 4`), then `*` (`3 * 4 = 12`), then `+` (`2 + 12 = 14`). The final result is `14`.

Now let us see a few examples.

Let's start with arithmetic operators.

In [ ]:
a = 10
b = 3

print(a + b)
print(a - b)
print(a * b)
print(a / b)
print(a // b)
print(a % b)
print(a ** b)

13
7
30
3.3333333333333335
3
1
1000


Here `a = 10` and `b = 3` are the operands. In order, the outputs are the results of addition, subtraction, multiplication, division, floor division, modulus, and exponent.

Notice how `/` gives us a decimal value `3.333...` while `//` rounds it down to `3`. The `%` operator gives us the remainder after division, which is `1` here since `10 = 3*3 + 1`.

Now let's see the assignment operators. These are shortcuts that combine an arithmetic operation with assignment.

In [ ]:
x = 10
print(x)

x += 5
print(x)

x -= 3
print(x)

x *= 2
print(x)

x //= 4
print(x)

10
15
12
24
6


Each line updates `x` based on its previous value and prints it right after.

- `x = 10` sets the starting value.
- `x += 5` is the same as `x = x + 5`, so `x` becomes `15`.
- `x -= 3` is the same as `x = x - 3`, so `x` becomes `12`.
- `x *= 2` is the same as `x = x * 2`, so `x` becomes `24`.
- `x //= 4` is the same as `x = x // 4`, so `x` becomes `6`.

This shorthand becomes very useful when we want to update a variable's value repeatedly, especially inside loops which we will see soon.

Next, let's look at comparison operators. These always give us a Boolean result, either `True` or `False`.

In [ ]:
p = 15
q = 20

print(p == q)
print(p != q)
print(p > q)
print(p < q)
print(p >= 15)
print(q <= 20)

False
True
False
True
True
True


Since `p` holds `15` and `q` holds `20`, they are not equal, so `p == q` gives `False` and `p != q` gives `True`.

Since `p` is smaller than `q`, `p > q` is `False` and `p < q` is `True`. The last two lines check equality boundaries with `>=` and `<=`. Since `p` is exactly `15` and `q` is exactly `20`, both conditions match and give `True`.

Now, let's combine multiple conditions using logical operators.

In [ ]:
age = 22
has_id = True

print(age >= 18 and has_id)
print(age < 18 or has_id)
print(not has_id)

True
True
False


The `and` operator needs both sides to be `True` to give `True`. Here `age >= 18` is `True` and `has_id` is `True`, so the result is `True`.

The `or` operator only needs one side to be `True`. Even though `age < 18` is `False`, `has_id` is `True`, so the overall result is `True`. Finally, `not has_id` simply flips `True` to `False`.

Let's also see membership operators, which help us check if a value exists inside a sequence like a list.

In [ ]:
fruits = ["apple", "banana", "cherry"]

print("banana" in fruits)
print("mango" in fruits)
print("mango" not in fruits)

True
False
True


Here `fruits` is a list containing three string values. Since `"banana"` exists in the list, `"banana" in fruits` returns `True`.

But `"mango"` is not present, so `"mango" in fruits` returns `False`, and naturally `"mango" not in fruits` returns `True`.

Lastly, let's understand identity operators using `is` and `is not`. These check whether two variables point to the exact same object in memory, not just whether their values are equal.

In [ ]:
list1 = [1, 2, 3]
list2 = [1, 2, 3]
list3 = list1

print(list1 == list2)
print(list1 is list2)
print(list1 is list3)

True
False
True


`list1` and `list2` have the same values, so `list1 == list2` is `True`. But they are two separate objects created independently in memory, so `list1 is list2` is `False`.

On the other hand, `list3` was assigned directly from `list1`, meaning it points to the same object in memory, so `list1 is list3` gives `True`.

### Loops

Till now, whatever code we wrote executed line by line, once each. But in real life, we often need to repeat an action multiple times.

Suppose you want to print numbers from 1 to 100. Would you write 100 print statements? Certainly not. This is where **loops** come in.

A loop allows us to repeat a block of code multiple times without writing it again and again. Python provides two main types of loops:

- The `for` loop, used when we want to repeat an action a specific number of times or over a sequence of values.
- The `while` loop, used when we want to repeat an action as long as a certain condition remains true.

#### The `for` loop

Before jumping into code, let's understand a few important terms.

A `for` loop is generally used to go through (or **iterate** over) a sequence of values, one at a time. This sequence could be a list, a string, a range of numbers, or any other collection of values.

The variable used inside a `for` loop to hold each value during the iteration is called the **iterator variable** (also sometimes called the loop variable). It automatically takes each value from the sequence, one after another, until the sequence is exhausted.

The basic syntax looks like this:

```python
for iterator_variable in sequence:
    # block of code to repeat
```

Notice the colon `:` at the end of the `for` line, and the indentation of the code below it. This indentation tells Python which lines belong inside the loop. This is similar to how we saw indentation matter in conditional statements.

Often, we don't already have a ready sequence to loop over, but just want to repeat something a fixed number of times, like counting from 1 to 5. For this, Python gives us a built-in function called `range()`.

The `range()` function generates a sequence of numbers, which we can then iterate over using a `for` loop.

It can be used in three ways:

- `range(stop)`: generates numbers starting from `0` up to (but not including) `stop`.
- `range(start, stop)`: generates numbers starting from `start` up to (but not including) `stop`.
- `range(start, stop, step)`: generates numbers starting from `start` up to (but not including) `stop`, increasing (or decreasing) by `step` each time.

Let's understand all of this with examples now.

In [ ]:
for i in range(5):
  print(i)

0
1
2
3
4


Here, `range(5)` generates numbers from `0` to `4` (5 numbers in total, but not including 5). The variable `i` is our iterator variable — it takes the value `0` first, runs the loop body (`print(i)`), then takes `1`, runs again, and so on until it has taken the value `4`.

Notice that the sequence starts from `0` by default, not `1`.

In [ ]:
for i in range(2, 6):
  print(i)

2
3
4
5


This time we used `range(2, 6)`, which means the sequence starts from `2` and goes up to (but not including) `6`. So the values taken by `i` are `2, 3, 4, 5`.

In [ ]:
for i in range(1, 10, 2):
  print(i)

1
3
5
7
9


Here we added a third value, `2`, which is the **step**. Instead of increasing by `1` each time, `i` now increases by `2`. So starting from `1`, we get `1, 3, 5, 7, 9` — all odd numbers below `10`.

The step value can also be negative, which allows us to count backward.

In [ ]:
for i in range(10, 0, -2):
  print(i)

10
8
6
4
2


Here, `i` starts at `10` and decreases by `2` each time, stopping just before it reaches `0`. So the values printed are `10, 8, 6, 4, 2`.

The `for` loop is not limited to just numbers from `range()`. It can iterate over other sequences too, like a list or even a string, since a string is just a sequence of characters.

In [ ]:
fruits = ["apple", "banana", "cherry"]

for fruit in fruits:
  print(fruit)

apple
banana
cherry


Here, our iterator variable is named `fruit` (we can name it anything meaningful). It takes each value from the `fruits` list one by one — first `"apple"`, then `"banana"`, then `"cherry"` — and the loop body prints it each time.

In [ ]:
name = "Alice"

for letter in name:
  print(letter)

A
l
i
c
e


Since a string is a sequence of characters, the `for` loop goes through it one character at a time. So `letter` takes the value `A`, then `l`, then `i`, then `c`, then `e`, printing each on a new line.

#### The `while` loop

Unlike the `for` loop, which is generally used when we know how many times we want to repeat something (or already have a sequence to go through), the `while` loop is used when we want to repeat an action **as long as a condition remains true**.

The basic syntax looks like this:

```python
while condition:
  # block of code to repeat
```

The condition is checked before every repetition. If the condition is `True`, the loop body runs. Once the loop body finishes, the condition is checked again. This continues until the condition becomes `False`, at which point the loop stops.

It is important to make sure that something inside the loop eventually makes the condition `False`. Otherwise, the loop will run forever, known as an **infinite loop**.

Let's see this with examples.

In [ ]:
count = 1

while count <= 5:
  print(count)
  count += 1

1
2
3
4
5


Here, `count` starts at `1`. The condition `count <= 5` is checked before each run of the loop body.

As long as this condition is `True`, the loop prints the current value of `count` and then increases it by `1` using `count += 1`. Once `count` becomes `6`, the condition `count <= 5` becomes `False`, and the loop stops.

Notice that `count += 1` is essential here. Without it, `count` would always remain `1`, the condition would always stay `True`, and we would end up with an infinite loop.

In [ ]:
num = 20

while num > 0:
  print(num)
  num -= 5

20
15
10
5


This time, `num` starts at `20` and the loop runs as long as `num` is greater than `0`. Each time, we print `num` and then reduce it by `5` using `num -= 5`. So the values printed are `20, 15, 10, 5`. Once `num` becomes `0`, the condition `num > 0` becomes `False`, and the loop ends.

## Working with Collections of Data

Till now, we have mostly worked with single values stored in variables, like a number or a name. But real-world data rarely comes as a single value. We often deal with a collection of values together, like a list of student names, or a set of marks.

Python gives us several built-in data structures to store collections of values. We will look at four of them:

- **Strings** – a sequence of characters
- **Lists** – an ordered, changeable collection of values
- **Tuples** – an ordered, unchangeable collection of values
- **Dictionaries** – a collection of key-value pairs

All four of these are called **sequences** or **collections**, since they hold more than one value together. Before we explore each one individually, let's understand one function that works across all of them: the `len()` function.

### The `len()` function

Whenever we work with a collection of values, one of the most common things we want to know is: how many values does it hold?

The `len()` function (short for **length**) is a built-in Python function that returns the number of items present in a collection. It works on strings, lists, tuples, and dictionaries alike, though what it "counts" depends slightly on the structure, as we will see.

The syntax is simple:

```python
len(collection)
```

Let's see this in action across different types of collections, before we study each of them in detail.

In [ ]:
name = "Alice"
marks = [45, 78, 92, 34, 60]
coordinates = (10, 20, 30)
student = {"name": "Alice", "age": 20}

print(len(name))
print(len(marks))
print(len(coordinates))
print(len(student))

5
5
3
2


- `len(name)` gives `5`, since the string `"Alice"` has 5 characters.
- `len(marks)` gives `5`, since the list has 5 elements.
- `len(coordinates)` gives `3`, since the tuple has 3 elements.
- `len(student)` gives `2`, since the dictionary has 2 key-value pairs. Note that `len()` counts the number of *pairs*, not individual keys and values separately.

Keep the `len()` function in mind as we go through each structure below. We will use it often, especially while looping, to know exactly how many times we need to repeat an action.

### Strings

We have already used strings before, like `"Alice"` or `"Hello"`. A string is simply a sequence of characters enclosed in quotes, and it can represent anything from a single word to entire paragraphs of text.

Since a string is a sequence, each character inside it occupies a specific position, called an **index**. Indexing in Python starts from `0`, not `1`. So in the string `"Alice"`, the character `A` is at index `0`, `l` is at index `1`, and so on, up to `e` at index `4` (since the string has 5 characters in total, matching what `len()` told us earlier).

We can also use **negative indexing** to access characters counting from the end of the string. The last character is at index `-1`, the second-last at `-2`, and so on. This is especially handy when we do not know the exact length of a string but still want to access its last few characters.

Let's see this with an example.

In [ ]:
name = "Alice"

print(name[0])
print(name[4])
print(name[-1])

A
e
e


Here, `name[0]` gives us `A`, the first character. `name[4]` gives us `e`, the last character counted from the front. `name[-1]` also gives us `e`, but this time counted from the back. Since `len(name)` was `5`, the valid forward indices range from `0` to `4`.

We can also access a portion of a string using **slicing**. Slicing uses the format `string[start:stop]`, which gives us characters from index `start` up to (but not including) `stop`.

In [ ]:
word = "Programming"

print(word[0:4])
print(word[4:])
print(word[:4])

Prog
ramming
Prog


`word[0:4]` gives us characters from index `0` to `3`, which is `"Prog"`. `word[4:]` means starting from index `4` all the way to the end, giving `"ramming"`. `word[:4]` means starting from the beginning up to index `3`, giving `"Prog"` again, since leaving out the start defaults to `0`.

Since a string is a sequence, we can iterate over it character by character using a `for` loop. We can even combine this with `len()` and `range()` to loop using index positions instead of directly over characters, which is useful when we need to know the position of each character too.

In [ ]:
word = "Data"

for letter in word:
  print(letter)

for i in range(len(word)):
  print(i, word[i])

D
a
t
a
0 D
1 a
2 t
3 a


The first loop directly iterates over each character of `"Data"`. The second loop instead iterates over index numbers generated by `range(len(word))`, which gives `0, 1, 2, 3` since `len(word)` is `4`. Using `word[i]` inside the loop, we fetch the character at each index, along with the index itself. This pattern of using `range(len(...))` is very common when both the position and the value are needed.

One important thing to remember about strings is that they are **immutable**, meaning once a string is created, its individual characters cannot be changed directly. Let's actually try to do this and see what happens.

In [ ]:
name = "Alice"
name[0] = "M"

TypeError: 'str' object does not support item assignment

This gives us a `TypeError: 'str' object does not support item assignment`. This confirms that strings do not allow us to modify a character at a given index directly.

So how do string methods like `upper()` or `replace()` appear to "change" the string then? They don't actually modify the original string — they create and return a brand new string instead. Let's verify this.

In [ ]:
name = "Alice"
new_name = name.upper()

print(name)
print(new_name)

Alice
ALICE


Notice that `name` still holds `"Alice"`, completely unchanged. The `upper()` method created a new string `"ALICE"` and we stored it separately in `new_name`. This is the behavior of every string method — the original string is never touched.

Here is a table listing the most commonly used string methods, along with their syntax and arguments.

| Method | Syntax | Arguments | Description | Example |
|---|---|---|---|---|
| `upper()` | `string.upper()` | None | Converts all characters to uppercase | `"hi".upper()` → `"HI"` |
| `lower()` | `string.lower()` | None | Converts all characters to lowercase | `"HI".lower()` → `"hi"` |
| `capitalize()` | `string.capitalize()` | None | Makes only the first character uppercase, rest lowercase | `"hello world".capitalize()` → `"Hello world"` |
| `title()` | `string.title()` | None | Makes the first letter of every word uppercase | `"hello world".title()` → `"Hello World"` |
| `strip()` | `string.strip()` | None (optional: characters to remove) | Removes leading and trailing spaces (or given characters) | `"  hi  ".strip()` → `"hi"` |
| `replace()` | `string.replace(old, new)` | `old` (mandatory), `new` (mandatory) | Replaces all occurrences of `old` with `new` | `"hi bob".replace("bob", "sam")` → `"hi sam"` |
| `split()` | `string.split(separator)` | `separator` (optional, default is space) | Splits the string into a list of substrings | `"a,b,c".split(",")` → `['a', 'b', 'c']` |
| `join()` | `separator.join(iterable)` | `iterable` (mandatory) | Joins elements of a list/tuple into a string using the given separator | `"-".join(["a", "b"])` → `"a-b"` |
| `find()` | `string.find(value)` | `value` (mandatory) | Returns the index of first occurrence of `value`, or `-1` if not found | `"hello".find("l")` → `2` |
| `count()` | `string.count(value)` | `value` (mandatory) | Counts number of occurrences of `value` in the string | `"hello".count("l")` → `2` |
| `startswith()` | `string.startswith(value)` | `value` (mandatory) | Checks if string starts with `value`, returns Boolean | `"hello".startswith("he")` → `True` |
| `endswith()` | `string.endswith(value)` | `value` (mandatory) | Checks if string ends with `value`, returns Boolean | `"hello".endswith("lo")` → `True` |
| `isdigit()` | `string.isdigit()` | None | Checks if all characters are digits, returns Boolean | `"123".isdigit()` → `True` |
| `isalpha()` | `string.isalpha()` | None | Checks if all characters are alphabets, returns Boolean | `"abc".isalpha()` → `True` |

### Lists

A list is an ordered collection of values, and unlike a string, it can hold values of different datatypes together in a single collection. A list is defined using square brackets `[ ]`, with values separated by commas.

In [ ]:
marks = [45, 78, 92, 34, 60]
print(marks)
print(len(marks))

mixed = ["Alice", 25, True, 5.5]
print(mixed)

[45, 78, 92, 34, 60]
5
['Alice', 25, True, 5.5]


Here, `marks` is a list of integers, and `len(marks)` confirms it has `5` elements. The `mixed` list shows that a single list can hold a string, an integer, a boolean, and a float all together, something a string cannot do since it only holds characters.

Just like strings, lists also support indexing, negative indexing, and slicing, since a list is also a sequence.

In [ ]:
marks = [45, 78, 92, 34, 60]

print(marks[0])
print(marks[-1])
print(marks[1:4])

45
60
[78, 92, 34]


`marks[0]` gives the first value `45`. `marks[-1]` gives the last value `60`. `marks[1:4]` gives a new list containing values from index `1` to `3`, which is `[78, 92, 34]`.

Unlike strings, lists are **mutable**, meaning we can change their content after creation, without needing to create a new list. Let's actually try modifying a list directly and confirm this works, unlike what happened with strings.

In [ ]:
marks = [45, 78, 92, 34, 60]
marks[0] = 50

print(marks)

[50, 78, 92, 34, 60]


Here, we directly changed the value at index `0` from `45` to `50`, and the original `marks` list itself gets updated in place, unlike what we saw with string methods.

We can iterate over a list using a `for` loop, and just like with strings, we can also combine it with `len()` and `range()` when we need index positions.

In [ ]:
marks = [45, 78, 92, 34, 60]

for mark in marks:
  print(mark)

for i in range(len(marks)):
  print(i, marks[i])

45
78
92
34
60
0 45
1 78
2 92
3 34
4 60


The first loop goes through each value directly. The second loop goes through index positions from `0` to `len(marks) - 1`, letting us access both the position and the value at each step using `marks[i]`.

Here is a table listing the most commonly used list methods, along with their syntax and arguments.

| Method | Syntax | Arguments | Description | Example |
|---|---|---|---|---|
| `append()` | `list.append(value)` | `value` (mandatory) | Adds a single value at the end of the list | `[1,2].append(3)` → `[1, 2, 3]` |
| `insert()` | `list.insert(index, value)` | `index` (mandatory), `value` (mandatory) | Inserts `value` at the given `index` | `[1,3].insert(1, 2)` → `[1, 2, 3]` |
| `extend()` | `list.extend(iterable)` | `iterable` (mandatory) | Adds all elements of another list/iterable to the end | `[1,2].extend([3,4])` → `[1, 2, 3, 4]` |
| `remove()` | `list.remove(value)` | `value` (mandatory) | Removes the first occurrence of `value` | `[1,2,2].remove(2)` → `[1, 2]` |
| `pop()` | `list.pop(index)` | `index` (optional, default is last item) | Removes and returns the item at `index` | `[1,2,3].pop()` → `3`, list becomes `[1, 2]` |
| `sort()` | `list.sort(reverse=False)` | `reverse` (optional, default `False`) | Sorts the list in ascending order (or descending if `reverse=True`) | `[3,1,2].sort()` → `[1, 2, 3]` |
| `reverse()` | `list.reverse()` | None | Reverses the order of elements in the list | `[1,2,3].reverse()` → `[3, 2, 1]` |
| `index()` | `list.index(value)` | `value` (mandatory) | Returns the index of the first occurrence of `value` | `[1,2,3].index(2)` → `1` |
| `count()` | `list.count(value)` | `value` (mandatory) | Counts number of occurrences of `value` | `[1,2,2,3].count(2)` → `2` |
| `clear()` | `list.clear()` | None | Removes all elements, leaving an empty list | `[1,2,3].clear()` → `[]` |
| `copy()` | `list.copy()` | None | Returns a shallow copy of the list | `[1,2].copy()` → `[1, 2]` |

Let's try out a few of these directly.

In [ ]:
fruits = ["banana", "apple", "cherry"]
fruits.append("mango")
print(fruits)

['banana', 'apple', 'cherry', 'mango']


In [ ]:
numbers = [10, 30, 40]
numbers.insert(1, 20)
print(numbers)

[10, 20, 30, 40]


In [ ]:
fruits = ["banana", "apple"]
fruits.extend(["cherry", "mango"])
print(fruits)

['banana', 'apple', 'cherry', 'mango']


In [ ]:
fruits = ["banana", "apple", "cherry", "apple"]
fruits.remove("apple")
print(fruits)

['banana', 'cherry', 'apple']


In [ ]:
numbers = [10, 20, 30]
removed_value = numbers.pop()
print(numbers)
print(removed_value)

[10, 20]
30


In [ ]:
numbers = [30, 10, 20]
numbers.sort()
print(numbers)

[10, 20, 30]


In [ ]:
numbers = [1, 2, 3]
numbers.reverse()
print(numbers)

[3, 2, 1]


In [ ]:
fruits = ["banana", "apple", "cherry"]
print(fruits.index("cherry"))

2


In [ ]:
numbers = [1, 2, 2, 3, 2]
print(numbers.count(2))

3


In [ ]:
numbers = [1, 2, 3]
numbers.clear()
print(numbers)

[]


Each method above modifies the original list directly (except `pop()` and `index()`, which also return a value we can use), confirming once again that lists are mutable.

### Tuples

A tuple is an ordered collection of values, very similar to a list. Just like a list, it can hold values of different datatypes together, and each value has a fixed position (index).

The key difference is that a tuple is **immutable**, meaning once created, its values cannot be changed, added, or removed. A tuple is defined using round brackets `( )`, with values separated by commas.

Tuples are generally used when we want to store a fixed collection of values that should not change throughout the program, such as coordinates, days of the week, or fixed configuration values.

In [ ]:
coordinates = (10, 20)
print(coordinates)

(10, 20)


In [ ]:
info = ("Alice", 25, "Engineer")
print(info)
print(len(info))

('Alice', 25, 'Engineer')
3


Here, `coordinates` is a tuple holding two integer values, and `info` is a tuple holding a mix of a string, an integer, and another string. `len(info)` gives us `3`, confirming the tuple holds three values, just as `len()` worked for lists and strings.

A tuple with a single value needs a comma after it, otherwise Python will not treat it as a tuple. Let's see this.

In [ ]:
single_value = (5)
print(type(single_value))

single_tuple = (5,)
print(type(single_tuple))

<class 'int'>
<class 'tuple'>


Notice that `single_value` is actually just an `int`, since without a comma, Python treats `(5)` as a regular value inside parentheses, not a tuple. Adding a comma after `5` as in `(5,)` is what actually makes it a tuple.

Just like lists and strings, tuples support **indexing**, **negative indexing**, and **slicing**, since a tuple is also a sequence.

In [ ]:
info = ("Alice", 25, "Engineer")
print(info[0])
print(info[-1])
print(info[0:2])

Alice
Engineer
('Alice', 25)


`info[0]` gives `"Alice"`, the first value. `info[-1]` gives `"Engineer"`, the last value counted from the back. `info[0:2]` gives a new tuple `("Alice", 25)`, containing values from index `0` to `1`.

We can also iterate over a tuple using a `for` loop, since it is a sequence, just like we did with lists and strings.

In [ ]:
info = ("Alice", 25, "Engineer")

for item in info:
  print(item)

Alice
25
Engineer


Now let's turn our attention to the immutability of tuples. Let's actually try to modify a tuple and see what happens, just like we did with strings.

In [ ]:
info = ("Alice", 25, "Engineer")
info[1] = 26

TypeError: 'tuple' object does not support item assignment

This gives us a `TypeError: 'tuple' object does not support item assignment`. This confirms the core difference between a list and a tuple — a list allows this kind of direct modification, but a tuple does not.

This is exactly why tuples are preferred when we want to store a fixed collection of values that should never accidentally change during the program.

Since tuples cannot be modified, they only support two built-in methods.

| Method | Syntax | Arguments | Description | Example |
|---|---|---|---|---|
| `count()` | `tuple.count(value)` | `value` (mandatory) | Counts number of occurrences of `value` in the tuple | `(1,2,2,3).count(2)` → `2` |
| `index()` | `tuple.index(value)` | `value` (mandatory) | Returns the index of the first occurrence of `value` | `(1,2,3).index(3)` → `2` |

Let's try out both of these directly.

In [ ]:
numbers = (10, 20, 10, 30, 10)
print(numbers.count(10))

3


In [ ]:
numbers = (10, 20, 10, 30, 10)
print(numbers.index(30))

3


`count(10)` tells us how many times the value `10` appears in the tuple, which is `3`. `index(30)` tells us the position at which the value `30` first appears, which is index `3`.

### Dictionaries

A dictionary is different from strings, lists, and tuples. Instead of storing values at an ordered position (index), it stores data as **key-value pairs**. Each value is associated with a unique key, and we access the value using its key rather than a position.

A dictionary is defined using curly brackets `{ }`. Inside, each key and its corresponding value are written together separated by a colon `:`, and different key-value pairs are separated by commas.

The basic syntax looks like this:

```python
dictionary_name = {key1: value1, key2: value2}
```

Dictionaries are especially useful when we want to label our data meaningfully, such as connecting a person's name to their age, rather than just remembering which index holds what.

In [ ]:
student = {"name": "Alice", "age": 20, "course": "Data Science"}
print(student)

{'name': 'Alice', 'age': 20, 'course': 'Data Science'}


In [ ]:
student = {"name": "Alice", "age": 20, "course": "Data Science"}
print(len(student))

3


Here, `"name"`, `"age"`, and `"course"` are the **keys**, and `"Alice"`, `20`, and `"Data Science"` are their corresponding **values**. `len(student)` gives `3`, since it counts the number of key-value pairs, not the individual keys and values separately.

A key must always be unique within a dictionary, and is typically a string or a number. Values, on the other hand, can be of any datatype, and can even repeat.

Let's now see how to access a value using its key.

In [ ]:
student = {"name": "Alice", "age": 20, "course": "Data Science"}
print(student["name"])

Alice


In [ ]:
student = {"name": "Alice", "age": 20, "course": "Data Science"}
print(student["age"])

20


`student["name"]` gives us `"Alice"`, and `student["age"]` gives us `20`. Notice how we use square brackets `[ ]` with the key inside, similar to how we used index numbers with lists, except here we use a key instead of a position.

Let's see what happens when we try to access a key that doesn't exist.

In [ ]:
student = {"name": "Alice", "age": 20, "course": "Data Science"}
print(student["salary"])

KeyError: 'salary'

This gives us a `KeyError: 'salary'`, since no such key exists in the dictionary. This is different from a list's `IndexError`, but the idea is similar — trying to access something that isn't there results in an error.

Now let's understand mutability. Dictionaries are mutable, similar to lists. We can update an existing value or add a completely new key-value pair after the dictionary has been created.

In [ ]:
student = {"name": "Alice", "age": 20}
student["age"] = 21
print(student)

{'name': 'Alice', 'age': 21}


In [ ]:
student = {"name": "Alice", "age": 20}
student["grade"] = "A"
print(student)

{'name': 'Alice', 'age': 20, 'grade': 'A'}


The first cell updates the existing key `"age"` to a new value `21`. The second cell adds a brand new key `"grade"`, which did not exist before. Both changes happened directly on the original `student` dictionary, without errors, confirming that dictionaries are mutable, just like lists.

We can iterate over a dictionary using a `for` loop too, but by default, it iterates over the keys, not the values.

In [ ]:
student = {"name": "Alice", "age": 20, "course": "Data Science"}

for key in student:
  print(key, student[key])

name Alice
age 20
course Data Science


Here, `key` takes each key from the dictionary one at a time, and we use `student[key]` to fetch its corresponding value inside the loop, printing both together.

Here is a table listing the most commonly used dictionary methods, along with their syntax and arguments.

| Method | Syntax | Arguments | Description | Example |
|---|---|---|---|---|
| `keys()` | `dict.keys()` | None | Returns all the keys in the dictionary | `{"a":1,"b":2}.keys()` → `dict_keys(['a', 'b'])` |
| `values()` | `dict.values()` | None | Returns all the values in the dictionary | `{"a":1,"b":2}.values()` → `dict_values([1, 2])` |
| `items()` | `dict.items()` | None | Returns all key-value pairs together | `{"a":1}.items()` → `dict_items([('a', 1)])` |
| `get()` | `dict.get(key, default)` | `key` (mandatory), `default` (optional, default `None`) | Safely returns value for `key`; returns `default` instead of an error if key is missing | `{"a":1}.get("b", 0)` → `0` |
| `update()` | `dict.update(other_dict)` | `other_dict` (mandatory) | Updates the dictionary with key-value pairs from another dictionary | `{"a":1}.update({"b":2})` → `{"a":1, "b":2}` |
| `pop()` | `dict.pop(key)` | `key` (mandatory) | Removes the given `key` and returns its value | `{"a":1,"b":2}.pop("a")` → `1` |
| `popitem()` | `dict.popitem()` | None | Removes and returns the last inserted key-value pair | `{"a":1,"b":2}.popitem()` → `('b', 2)` |
| `clear()` | `dict.clear()` | None | Removes all key-value pairs, leaving an empty dictionary | `{"a":1}.clear()` → `{}` |
| `copy()` | `dict.copy()` | None | Returns a shallow copy of the dictionary | `{"a":1}.copy()` → `{"a": 1}` |

Let's try out each of these directly.

In [ ]:
student = {"name": "Alice", "age": 20}
print(student.keys())

dict_keys(['name', 'age'])


In [ ]:
student = {"name": "Alice", "age": 20}
print(student.values())

dict_values(['Alice', 20])


In [ ]:
student = {"name": "Alice", "age": 20}
print(student.items())

dict_items([('name', 'Alice'), ('age', 20)])


- `keys()` returns all the keys in the dictionary.
- `values()` returns all the values in the dictionary.
- `items()` returns each key-value pair together, useful when we want both the key and value at once during iteration.

In [ ]:
student = {"name": "Alice", "age": 20}
print(student.get("salary", 0))

0


Notice how `get("salary", 0)` returns `0` instead of raising a `KeyError`, since `"salary"` does not exist in the dictionary. This is why `get()` is often safer to use than direct key access when we are unsure whether a key exists.

In [ ]:
student = {"name": "Alice", "age": 20}
student.update({"grade": "A", "age": 21})
print(student)

{'name': 'Alice', 'age': 21, 'grade': 'A'}


In [ ]:
student = {"name": "Alice", "age": 20}
removed_value = student.pop("age")
print(student)
print(removed_value)

{'name': 'Alice'}
20


In [ ]:
student = {"name": "Alice", "age": 20}
last_item = student.popitem()
print(student)
print(last_item)

{'name': 'Alice'}
('age', 20)


In [ ]:
student = {"name": "Alice", "age": 20}
student.clear()
print(student)

{}


### Comparing Strings, Lists, Tuples, and Dictionaries — In Detail

We have now studied all four data structures individually. Let's bring everything together into one detailed comparison, covering their syntax, behavior, and typical use cases.

#### Similarities

- All four are used to store collections of data rather than a single value.
- All four support the `len()` function to find out how many items they hold.
- Strings, Lists, and Tuples are all **ordered sequences**, meaning each item has a fixed position, and Dictionaries (from Python 3.7 onwards) also maintain the order in which items were inserted.
- Strings, Lists, and Tuples all support indexing, negative indexing, and slicing, since they are sequences.
- All four can be looped over using a `for` loop.
- All four can be checked for membership using the `in` and `not in` operators.

#### Detailed Differences

| Feature | String | List | Tuple | Dictionary |
|---|---|---|---|---|
| Symbol used | Quotes `" "` or `' '` | Square brackets `[ ]` | Round brackets `( )` | Curly brackets `{ }` |
| Stores | Only characters | Any datatype, mixed | Any datatype, mixed | Key-value pairs (values can be any datatype) |
| Ordered | Yes | Yes | Yes | Yes (insertion order preserved) |
| Mutable (changeable) | No | Yes | No | Yes |
| Access method | Index (position) | Index (position) | Index (position) | Key (label) |
| Supports negative indexing | Yes | Yes | Yes | No (keys are not numeric positions) |
| Supports slicing | Yes | Yes | Yes | No |
| What `len()` counts | Number of characters | Number of elements | Number of elements | Number of key-value pairs |
| Allows duplicate values | Yes | Yes | Yes | Keys must be unique; values can repeat |
| Can hold mixed datatypes | No (only characters) | Yes | Yes | Yes (for values; keys are usually string/number) |
| Can grow or shrink in size | No | Yes (append, remove, etc.) | No | Yes (add, update, remove keys) |
| Error on invalid access | `IndexError` | `IndexError` | `IndexError` | `KeyError` |
| Error on modification attempt | `TypeError` | Not applicable (allowed) | `TypeError` | Not applicable (allowed) |
| Number of built-in methods | Many (text-related) | Many (general-purpose) | Very few (`count`, `index`) | Moderate (key/value focused) |
| Common use case | Handling and processing text | General-purpose, changeable collection of items | Fixed collection that should not change | Structured or labeled data, like a record |
| Real-world example | A person's name or a sentence | A shopping list that keeps updating | Days of the week, GPS coordinates | A student's profile (name, age, marks) |

#### Quick Memory Tip

- **Strings and Tuples**: think of them as "sealed" — once written, they cannot be edited.
- **Lists and Dictionaries**: think of them as "open" — they can be edited, added to, or shrunk anytime.
- **Strings, Lists, Tuples**: think "position matters" — accessed using index numbers.
- **Dictionaries**: think "label matters" — accessed using keys, not position.